research/delf/delf/python/training/global_features/train_utils.py

```diff
@@ -0,0 +1,368 @@
+              'batch_time.val:.3f} '
+              '(Batch Time avg: {batch_time.avg:.3f})\t Data {'
+              'data_time.val:.3f} ('
+              'Time avg: {data_time.avg:.3f})\t Immediate loss value: {'
+              'loss.val:.4f} '
+              '(Loss avg: {loss.avg:.4f})'.format(
+                      state, epoch, batch_num + 1, all_batch_num,
+                      batch_time=batch_time,
+                      data_time=data_time, loss=losses), debug=True, log=False)
+    batch_num += 1
+
+  return losses.avg
+
+
+def test(datasets, net, epoch, writer=None, model_directory=None,
+         precompute_whitening=None, data_root='data', multiscale=[1.]):
+  """Testing step.
+
+  Evaluates the network on the provided test datasets by computing single-scale
+  mAP for easy/medium/hard cases. If `writer` is specified, saves the mAP
+  values in a tensorboard supported format.
+
+  Args:
+    datasets: List of dataset names for model testing (from
+      `_TEST_DATASET_NAMES`).
+    net: Network to evaluate.
+    epoch: Integer, epoch number.
+    writer: Tensorboard writer.
+    model_directory: String, path to the model directory.
+    precompute_whitening: Dataset used to learn whitening. If no
+      precomputation required, then `None`.
+    data_root: Absolute path to the data folder.
+    multiscale: List of scales for multiscale testing.
+  """
+  global_features_utils.debug_and_log(">> Testing step:")
+  global_features_utils.debug_and_log(
+          '>> Evaluating network on test datasets...')
+
+  # For testing we use image size of max 1024.
+  image_size = 1024
+
+  # Precompute whitening.
+  if precompute_whitening is not None:
+
+    # If whitening already precomputed, load it and skip the computations.
+    filename = os.path.join(
+            model_directory, 'learned_whitening_mP_{}_epoch.pkl'.format(epoch))
+    filename_layer = os.path.join(
+            model_directory,
+            'learned_whitening_layer_config_{}_epoch.pkl'.format(
+                    epoch))
+
+    if os.path.isfile(filename):
+      global_features_utils.debug_and_log(
+              '>> {}: Whitening for this epoch is already precomputed. '
+              'Loading...'.format(precompute_whitening))
+      with tf.io.gfile.GFile(filename, 'rb') as learned_whitening_file:
+        learned_whitening = pickle.load(learned_whitening_file)
+
+    else:
+      start = time.time()
+      global_features_utils.debug_and_log(
+              '>> {}: Learning whitening...'.format(precompute_whitening))
+
+      # Loading db.
+      db_root = os.path.join(data_root, 'train', precompute_whitening)
+      ims_root = os.path.join(db_root, 'ims')
+      db_fn = os.path.join(db_root,
+                           '{}-whiten.pkl'.format(precompute_whitening))
+      with tf.io.gfile.GFile(db_fn, 'rb') as f:
+        db = pickle.load(f)
+      images = [sfm120k.id2filename(db['cids'][i], ims_root) for i in
```

Comment:
I see that this seems to be specific to the sfm120k datasets. So let's mention it in the docstring that only such whitening datasets are supported?